project/src/Agents/GSTSpatialAgent.gd

```diff
@@ -0,0 +1,166 @@
+	if apply_linear_drag:
+		velocity = velocity.linear_interpolate(Vector3.ZERO, linear_drag_percentage)
+	velocity = body.move_and_slide(velocity)
+	if calculate_velocities:
+		linear_velocity = velocity
+
+
+func _apply_collide_steering(accel: Vector3, delta: float) -> void:
+	var velocity := GSTUtils.clampedv3(linear_velocity + accel, linear_speed_max)
+	if apply_linear_drag:
+		velocity = velocity.linear_interpolate(
+				Vector3.ZERO,
+				linear_drag_percentage
+		)
+	body.move_and_collide(velocity * delta)
+	if calculate_velocities:
+		linear_velocity = velocity
+
+
+func _apply_normal_steering(accel: Vector3, delta: float) -> void:
+	var velocity := GSTUtils.clampedv3(linear_velocity + accel, linear_speed_max)
+	if apply_linear_drag:
+		velocity = velocity.linear_interpolate(
+				Vector3.ZERO,
+				linear_drag_percentage
+		)
+	body.global_position += velocity * delta
+	if calculate_velocities:
+		linear_velocity = velocity
+
+
+func _apply_orientation_steering(angular_acceleration: float, delta: float) -> void:
+	var velocity = angular_velocity + angular_acceleration
+	if apply_angular_drag:
+		velocity = lerp(velocity, 0, angular_drag_percentage)
+	body.rotation += velocity * delta
+	if calculate_velocities:
+		angular_velocity = velocity
+
+
+func _set_use_physics(value: bool) -> void:
+	if use_physics and not value:
+		body.get_tree().disconnect("idle_frame", self, "_on_SceneTree_frame")
+		body.get_tree().connect("physics_frame", self, "_on_SceneTree_frame")
+	elif not use_physics and value:
+		body.get_tree().disconnect("physics_frame", self, "_on_SceneTree_frame")
+		body.get_tree().connect("idle_frame", self, "_on_SceneTree_frame")
+	use_physics = value
+
+
+func _set_body(value: Spatial) -> void:
+	body = value
+	if body is RigidBody:
+		_body_type = BodyType.RIGID
+	elif body is KinematicBody:
+		_body_type = BodyType.KINEMATIC
+	else:
+		_body_type = BodyType.NODE
+	
+	if _body_type == BodyType.RIGID:
+		linear_velocity = body.linear_velocity
+		angular_velocity = body.angular_velocity
+	else:
+		_last_position = body.global_position
+		_last_orientation = body.rotation.y
+	
+	position = _last_position
+	orientation = _last_orientation
+
+
+func _on_body_ready() -> void:
+	body.get_tree().connect("physics_frame", self, "_on_SceneTree_frame")
+	body.disconnect("ready", self, "_on_body_ready")
+	_set_body(body)
+
+
+func _on_SceneTree_frame() -> void:
+	var current_position: Vector3 = body.global_position
+	var current_orientation: float = body.rotation.y
+	
+	position = current_position
+	orientation = current_orientation
+	
+	if calculate_velocities:
+		if _applied_steering:
+			_applied_steering = false
+		else:
+			match _body_type:
+				BodyType.RIGID:
+					linear_velocity = GSTUtils.to_vector3(body.linear_velocity)
+					angular_velocity = body.angular_velocity
+				_:
+					linear_velocity = GSTUtils.clampedv3(
+							_last_position - current_position,
```

Comment:
Use one extra tab for hanging indentation. That's the standard with python at least (one indent level, so 4 spaces or 1 tab)

Reply:
The GDScript guidelines on the Godot docs [says 2 for hanging indentation](https://docs.godotengine.org/en/3.2/getting_started/scripting/gdscript/gdscript_styleguide.html#indentation) so that's what I have been going for.

Unless there's a difference between starting from a parenthesis VS code.

Reply:
Ah I see, all good then